manticore/native/models.py

```diff
@@ -140,3 +142,123 @@ def strlen(state, s):
+        return not Z3Solver.instance().can_be_true(constrs, byte == 0)
+    else:
+        return byte != 0
+
+
+def can_be_NULL(byte, constrs) -> bool:
+    """
+    Checks if a given byte read from memory can be NULL
+
+    :param byte: byte read from memory to be examined
+    :param constrs: state constraints
+    :return: whether a given byte is NULL or can be NULL
+    """
+    if issymbolic(byte):
+        return Z3Solver.instance().can_be_true(constrs, byte == 0)
+    else:
+        return byte == 0
+
+
+def strcpy(state: State, dst: Union[int, BitVec], src: Union[int, BitVec]) -> Union[int, BitVec]:
+    """
+    strcpy symbolic model
+
+    Algorithm: Copy every byte from the src to dst until finding a byte that can be or is NULL.
+    If the byte is NULL or is constrained to only the NULL value, append the NULL value to dst
+    and return. If the value can be NULL or another value write an `Expression` for every following
+    byte that sets a value to the src or dst byte according to the preceding bytes until a NULL
+    byte or effectively NULL byte is found.
+
+    :param state: current program state
+    :param dst: destination string address
+    :param src: source string address
+    :return: pointer to the dst
+    """
+    if issymbolic(src):
+        raise ConcretizeArgument(state.cpu, 1)
+
+    if issymbolic(dst):
+        raise ConcretizeArgument(state.cpu, 2)
+
+    cpu = state.cpu
+    constrs = state.constraints
+    ret = dst
+
+    # Copy until '\000' is reached or symbolic memory that can be '\000'
+    c = cpu.read_int(src, 8)
+    while cant_be_NULL(c, constrs):
+        cpu.write_int(dst, c, 8)
+        src += 1
+        dst += 1
+        c = cpu.read_int(src, 8)
+
+    # If the byte is symbolic and constrained to '\000' or is '\000' write concrete val and return
+    if is_NULL(c, constrs):
+        cpu.write_int(dst, 0, 8)
+        return ret
+
+    offset = 0
+    zeros: List[int] = []
+    src_val = cpu.read_int(src, 8)
+    while not is_NULL(src_val, constrs):
+        if can_be_NULL(c, constrs):
+            # If a byte can be NULL set the src_val for NULL, build the ITE, & add to the list of nulls
+            src_val = ITEBV(8, src_val != 0, src_val, 0)
+            _build_ITE(zeros, cpu, src, dst, offset, src_val)
+            zeros.append(offset)
+        else:
+            # If it can't be NULL just build the ITE
+            _build_ITE(zeros, cpu, src, dst, offset, src_val)
+        offset += 1
+        src_val = cpu.read_int(src + offset, 8)
+
+    # Build ITE Tree for NULL byte
+    src_val = 0
+    _build_ITE(zeros, cpu, src, dst, offset, src_val)
+
+    return ret
+
+
+def _build_ITE(
+    zeros: List[int],
+    cpu: Cpu,
+    src: Union[int, BitVec],
+    dst: Union[int, BitVec],
+    offset: int,
+    src_val: Union[int, BitVec],
+) -> None:
+    """
+    Builds ITE tree for each symbolic dst byte
+    """
+    dst_val = cpu.read_int(dst + offset, 8)
+    for zero in reversed(zeros):
```

Comment:
Suggestion: if you used a `collections.deque` (or `typing.Deque[int]`) instead of a list, you could avoid having to reverse the list every time, as you could add or remove in amortized constant time from either end with a `deque`.